File: frontend/src/utils/leagueSportRules.ts
// League sport classification helpers.
//
// These rules centralize the logic for "does this sport behave like a
// team-based league?" so LeagueSearchPage and LeagueDetailPage stay aligned.

/**
 * Returns true for motorsports (F1, NASCAR, WRC, ...).
 * Motorsports have no meaningful home/away team - all participants race
 * in every event, so the league is always considered monitored.
 */
export function isMotorsport(sport: string): boolean {
  const motorsports = [
    'Motorsport', 'Racing', 'Formula 1', 'F1', 'NASCAR', 'IndyCar',
    'MotoGP', 'WEC', 'Formula E', 'Rally', 'WRC', 'DTM', 'Super GT',
    'IMSA', 'V8 Supercars', 'Supercars', 'Le Mans',
  ];
  return motorsports.some(s => sport.toLowerCase().includes(s.toLowerCase()));
}

/**
 * Returns true for golf.
 * Golf tournaments have all players competing together - no home/away teams.
 */
export function isGolf(sport: string): boolean {
  return sport.toLowerCase() === 'golf';
}

/**
 * Returns true for individual-format tennis leagues (ATP/WTA tours).
 * Individual tours have no meaningful team data - all events should sync.
 * Team-based competitions (Fed Cup, Davis Cup, Olympics) return false.
 */
export function isIndividualTennis(sport: string, leagueName: string): boolean {
  if (sport.toLowerCase() !== 'tennis') return false;
  const nameLower = leagueName.toLowerCase();
  const individualTours = ['atp', 'wta'];
  const teamBased = ['fed cup', 'davis cup', 'olympic', 'billie jean king'];
  if (teamBased.some(t => nameLower.includes(t))) return false;
  return individualTours.some(t => nameLower.includes(t));
}

/**
 * Returns true for fighting sports (UFC, Boxing, MMA, Wrestling, etc.).
 * Fighting events use multi-part structure (Early Prelims, Prelims, Main Card).
 */
export function isFightingSport(sport: string): boolean {
  // 'Combat' is the sportarr-hub canonical sport name; TheSportsDB labels
  // the same sport 'Fighting'. Both must classify so MMA leagues like
  // MVP MMA (hub sport=Combat) skip the team-based event filter that
  // would otherwise drop fight events (TSDB doesn't populate home/away).
  // Keep aligned with the backend EventPartDetector.IsFightingSport list.
  const fightingSports = ['Fighting', 'Combat', 'MMA', 'UFC', 'Boxing', 'Kickboxing', 'Muay Thai', 'Wrestling'];
  return fightingSports.some(s => sport.toLowerCase().includes(s.toLowerCase()));
}

/**
 * Part options for a sport. Only fighting sports have multi-part episodes
 * (Early Prelims / Prelims / Main Card / Post Show); motorsports and other
 * sports have none. Order matches PartNumber on the backend
 * (EventPartDetector.CardSegment). Post Show only exists on PPV-style
 * events but is listed so it can be opted out at the league level (Fight
 * Night events ignore it because their per-event partStatuses omit it).
 */
export function getPartOptions(sport: string): string[] {
  if (isFightingSport(sport)) {
    return ['Early Prelims', 'Prelims', 'Main Card', 'Post Show'];
  }
  return [];
}

/**
 * Returns true for individual-player racket/cue sports (Badminton, Table Tennis, Snooker).
 * Tournaments feature individual players, not teams - all events should sync without team filtering.
 */
export function isIndividualRacketOrCueSport(sport: string): boolean {
  const s = sport.toLowerCase();
  return s === 'badminton' || s === 'table tennis' || s === 'snooker';
}

/**
 * Darts matches are between individual players, not teams.
 */
export function isDarts(sport: string): boolean {
  return sport.toLowerCase() === 'darts';
}

/**
 * Climbing competitions are individual climbers, not teams.
 */
export function isClimbing(sport: string): boolean {
  return sport.toLowerCase() === 'climbing';
}

/**
 * Gambling (Poker, WSOP) events are individual players in tournaments, not teams.
 */
export function isGambling(sport: string): boolean {
  return sport.toLowerCase() === 'gambling';
}

/**
 * Returns true for sports/leagues that have no meaningful home/away team structure.
 * These leagues should auto-monitor on add (no team selection required) and must
 * stay in sync with the backend's sportsWithoutTeamFiltering list in
 * LeagueEventSyncService.cs.
 */
export function isTeamlessSport(sport: string, leagueName: string): boolean {
  return (
    isMotorsport(sport) ||
    isGolf(sport) ||
    isDarts(sport) ||
    isClimbing(sport) ||
    isGambling(sport) ||
    isIndividualRacketOrCueSport(sport) ||
    isIndividualTennis(sport, leagueName)
  );
}

/**
 * Returns true for fighting leagues that monitor by event type (PPV, Fight Night,
 * Premium Live Event, ...) instead of by team. The edit modal hides the team
 * picker for these leagues, so they must be treated as teamless when computing
 * `monitored` on save — otherwise an empty `monitoredTeamIds` would force the
 * league off and override the user's event-type selection.
 */
export function usesFightingEventTypes(sport: string, leagueName: string): boolean {
  if (!isFightingSport(sport)) return false;
  const name = leagueName.toLowerCase();
  return (
    name.includes('ufc') ||
    name.includes('ultimate fighting') ||
    name.includes('wwe') ||
    name.includes('aew') ||
    name.includes('wrestling') ||
    name === 'one' ||
    name.includes('one championship') ||
    name.includes('one fc')
  );
}

File: src/Services/LeagueSportRules.cs
namespace Sportarr.Api.Services;

/// <summary>
/// Single source of truth for "which sports have no home/away team structure".
/// These leagues auto-monitor on add (no team selection required) and bypass
/// team-based event filtering. Must stay in sync with the frontend helper
/// isTeamlessSport in frontend/src/utils/leagueSportRules.ts.
/// </summary>
public static class LeagueSportRules
{
    private static readonly string[] TeamlessSports = new[]
    {
        // "Combat" is the hub canonical name for what TheSportsDB calls
        // "Fighting" — both must classify as teamless or fight events get
        // filtered to zero by the home/away team filter (TSDB never
        // populates home/away on fight events; what looks like a "team"
        // for an MMA promotion is actually a weight class, used for event
        // tagging not event filtering).
        "Fighting", "Combat",
        // "Racing" is the hub canonical name for what TheSportsDB calls
        // "Motorsport" — same divergence as Combat/Fighting above. Both must
        // classify as teamless: motorsport events never carry home/away teams,
        // so without this a freshly added F1/MotoGP/Formula E league is treated
        // as a team sport with no teams selected, left unmonitored, and never
        // synced (the UI sits on "Syncing events..." forever). The frontend
        // mirror already lists both spellings.
        "Cycling", "Motorsport", "Racing", "Golf", "Darts",
        "Climbing", "Gambling", "Badminton", "Table Tennis", "Snooker"
    };

    /// <summary>
    /// True for motorsport leagues regardless of which sport spelling upstream
    /// ships ("Motorsport" from TheSportsDB, "Racing" from the hub). Use this
    /// for session-type filtering (Race / Qualifying / Practice) instead of
    /// comparing against a single literal.
    /// </summary>
    public static bool IsMotorsport(string? sport)
    {
        if (string.IsNullOrEmpty(sport)) return false;
        return sport.Equals("Motorsport", System.StringComparison.OrdinalIgnoreCase)
            || sport.Equals("Racing", System.StringComparison.OrdinalIgnoreCase);
    }

    /// <summary>
    /// Sports that mean the same thing under different names. TheSportsDB, the
    /// hub and release naming each pick a different word for the same sport, so
    /// a straight string compare reports a mismatch between a league and its own
    /// events. Every group here is one sport.
    /// </summary>
    private static readonly string[][] EquivalentSports = new[]
    {
        new[] { "Fighting", "Combat", "MMA", "Mixed Martial Arts" },
        new[] { "Motorsport", "Racing", "Motorsports", "Auto Racing" },
        new[] { "American Football", "Football", "Gridiron" },
        new[] { "Ice Hockey", "Hockey" },
        new[] { "Association Football", "Soccer" },
    };

    /// <summary>
    /// The hub's canonical sport names, mirroring SPORT_SYNONYMS in the hub's
    /// sync_pipeline. One vocabulary across both apps means no translation
    /// layer between them. Combat is the umbrella sport; UFC, Bellator, ONE,
    /// Boxing and the wrestling promotions are leagues inside it.
    /// </summary>
    private static readonly Dictionary<string, string> CanonicalSportNames =
        new(System.StringComparer.OrdinalIgnoreCase)
        {
            ["football"] = "American Football",
            ["gridiron"] = "American Football",
            ["hockey"] = "Ice Hockey",
            ["racing"] = "Motorsport",
            ["motorsports"] = "Motorsport",
            ["auto racing"] = "Motorsport",
            ["mma"] = "Combat",
            ["mixed martial arts"] = "Combat",
            ["fighting"] = "Combat",
            ["boxing"] = "Combat",
            ["wrestling"] = "Combat",
            ["association football"] = "Soccer",
        };

    /// <summary>
    /// Returns the hub's canonical name for a sport, or the input unchanged
    /// when it is already canonical or unknown.
    /// </summary>
    public static string? CanonicalSport(string? sport)
    {
        if (string.IsNullOrWhiteSpace(sport)) return sport;
        return CanonicalSportNames.TryGetValue(sport.Trim(), out var canonical) ? canonical : sport;
    }

    /// <summary>
    /// True when two sport names describe the same sport. Use this instead of
    /// comparing the strings, or a correctly named release is judged against
    /// its own event as if it came from another sport.
    /// </summary>
    public static bool AreEquivalentSports(string? a, string? b)
    {
        if (string.IsNullOrWhiteSpace(a) || string.IsNullOrWhiteSpace(b)) return false;
        if (a.Equals(b, System.StringComparison.OrdinalIgnoreCase)) return true;

        foreach (var group in EquivalentSports)
        {
            var hasA = false;
            var hasB = false;
            foreach (var name in group)
            {
                if (a.Equals(name, System.StringComparison.OrdinalIgnoreCase)) hasA = true;
                if (b.Equals(name, System.StringComparison.OrdinalIgnoreCase)) hasB = true;
            }
            if (hasA && hasB) return true;
        }
        return false;
    }

    /// <summary>
    /// Returns true for sports/leagues that do not have meaningful home/away
    /// teams. Individual tennis tours (ATP/WTA) also qualify, but team-based
    /// tennis competitions (Fed Cup, Davis Cup, Olympics, Billie Jean King Cup)
    /// do not.
    /// </summary>
    public static bool IsTeamlessSport(string? sport, string? leagueName)
    {
        if (string.IsNullOrEmpty(sport)) return false;
        if (TeamlessSports.Contains(sport, System.StringComparer.OrdinalIgnoreCase)) return true;
        return IsIndividualTennisLeague(sport, leagueName ?? string.Empty);
    }

    public static bool IsIndividualTennisLeague(string sport, string leagueName)
    {
        if (!sport.Equals("Tennis", System.StringComparison.OrdinalIgnoreCase)) return false;
        var nameLower = leagueName.ToLowerInvariant();
        var teamBased = new[] { "fed cup", "davis cup", "olympic", "billie jean king" };
        if (teamBased.Any(t => nameLower.Contains(t))) return false;
        var individualTours = new[] { "atp", "wta" };
        return individualTours.Any(t => nameLower.Contains(t));
    }
}
